File: scripts/_sdk_e2e_client.mjs
// SDK e2e (Node) — runs with the TARBALL-INSTALLED astra-sdk from a /tmp project.
//
// Pulls the deployed artifact, serves it locally with LocalRunner, fires normal
// traffic then distribution-shifted traffic, and flushes telemetry. No repo
// imports — only the installed package + onnxruntime-node. Mirrors the Python
// _sdk_e2e_client.py so the shared _sdk_e2e_assert.py works unchanged.

import { readFileSync } from "node:fs";
import { parseArgs } from "node:util";

const { values } = parseArgs({ options: { handoff: { type: "string" } } });
const h = JSON.parse(readFileSync(values.handoff, "utf8"));

const { LocalRunner } = await import("astra-sdk");

// Prove we imported the INSTALLED tarball, not the repo source.
const resolved = import.meta.resolve("astra-sdk");
if (!resolved.includes("/node_modules/astra-sdk/")) {
  throw new Error(`must import the INSTALLED tarball, got ${resolved}`);
}
console.log(`   using ${resolved}`);

const runner = await LocalRunner.fromDeployment({
  baseUrl: h.baseUrl,
  deploymentId: h.deploymentId,
  apiKey: h.apiKey,
  cacheDir: "/tmp/astra-node-cache",
});

const spec = runner.inputSpecs()[0];
const name = spec.name;
const dims = (spec.dims ?? [1]).map((d) => (Number.isInteger(d) && d > 0 ? d : 1));
const size = dims.reduce((a, b) => a * b, 1);

function gauss() {
  let u = 0;
  let v = 0;
  while (u === 0) u = Math.random();
  while (v === 0) v = Math.random();
  return Math.sqrt(-2 * Math.log(u)) * Math.cos(2 * Math.PI * v);
}
function probe(shift = 0) {
  const data = new Float32Array(size);
  for (let i = 0; i < size; i++) data[i] = gauss() + shift;
  return { [name]: { data, dims, type: "float32" } };
}

const t0 = Date.now();
for (let i = 0; i < 100; i++) await runner.run(probe());
console.log(`   100 baseline inferences in ${((Date.now() - t0) / 1000).toFixed(1)}s`);

// Distribution shift: +5 sigma on every input feature.
for (let i = 0; i < 60; i++) await runner.run(probe(5.0));
console.log("   60 shifted inferences (input mean +5.0)");

await runner.close(); // final telemetry flush (beforeExit budget)
console.log("   telemetry flushed");

File: scripts/_sdk_e2e_client.py
#!/usr/bin/env python3
"""SDK e2e step 4 — runs with the WHEEL-INSTALLED astra_sdk from /tmp.

Pulls the deployed artifact, serves it locally with LocalRunner, fires
normal traffic then distribution-shifted traffic, and flushes telemetry.
No repo imports — only the installed package + numpy.
"""

from __future__ import annotations

import argparse
import json
import time


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--handoff", required=True)
    args = ap.parse_args()
    with open(args.handoff) as f:
        h = json.load(f)

    import numpy as np

    import astra_sdk
    from astra_sdk import LocalRunner

    assert not astra_sdk.__file__.startswith("/Users/kwonminjae/Desktop/Astra"), (
        f"must import the INSTALLED wheel, got {astra_sdk.__file__}")
    print(f"   using {astra_sdk.__file__}")

    runner = LocalRunner.from_deployment(
        h["deploymentId"], h["apiKey"], base_url=h["baseUrl"],
        cache_dir="/tmp/astra-sdk-cache",
    )
    print(f"   artifact cached at {runner.model_path}")

    meta = runner._input_meta  # noqa: SLF001 — e2e introspection
    name = meta[0].name
    shape = [d if isinstance(d, int) and d > 0 else 1 for d in meta[0].shape]
    rng = np.random.default_rng(7)

    t0 = time.time()
    for i in range(100):
        runner.run({name: rng.standard_normal(shape).astype(np.float32)})
    print(f"   100 baseline inferences in {time.time() - t0:.1f}s")

    # Distribution shift: +5 sigma on every input feature.
    for i in range(60):
        x = rng.standard_normal(shape).astype(np.float32) + 5.0
        runner.run({name: x})
    print("   60 shifted inferences (input mean +5.0)")

    runner.close()  # final telemetry flush (atexit budget)
    print("   telemetry flushed")


if __name__ == "__main__":
    main()
